File: packages/typescript/src/media/mime.ts
/**
 * MIME type detection and media utilities for Eliza.
 *
 * Provides robust MIME type detection from file buffers, headers, and extensions.
 */

// Lazy-loaded file-type module for MIME sniffing
type FileTypeResult = { ext: string; mime: string } | undefined;
type FileTypeFromBuffer = (
	buffer: ArrayBuffer | Uint8Array,
) => Promise<FileTypeResult>;

let fileTypeModule:
	| { fileTypeFromBuffer: FileTypeFromBuffer }
	| null
	| undefined;

async function getFileTypeFromBuffer(): Promise<FileTypeFromBuffer | null> {
	if (fileTypeModule === undefined) {
		try {
			fileTypeModule = await import("file-type");
		} catch {
			fileTypeModule = null;
		}
	}
	return fileTypeModule?.fileTypeFromBuffer ?? null;
}

/** Media kind categories */
export type MediaKind = "image" | "audio" | "video" | "document" | "unknown";

/** Map common MIME types to preferred file extensions */
const EXT_BY_MIME: Record<string, string> = {
	"image/heic": ".heic",
	"image/heif": ".heif",
	"image/jpeg": ".jpg",
	"image/png": ".png",
	"image/webp": ".webp",
	"image/gif": ".gif",
	"audio/ogg": ".ogg",
	"audio/mpeg": ".mp3",
	"audio/x-m4a": ".m4a",
	"audio/mp4": ".m4a",
	"video/mp4": ".mp4",
	"video/quicktime": ".mov",
	"application/pdf": ".pdf",
	"application/json": ".json",
	"application/zip": ".zip",
	"application/gzip": ".gz",
	"application/x-tar": ".tar",
	"application/x-7z-compressed": ".7z",
	"application/vnd.rar": ".rar",
	"application/msword": ".doc",
	"application/vnd.ms-excel": ".xls",
	"application/vnd.ms-powerpoint": ".ppt",
	"application/vnd.openxmlformats-officedocument.wordprocessingml.document":
		".docx",
	"application/vnd.openxmlformats-officedocument.spreadsheetml.sheet": ".xlsx",
	"application/vnd.openxmlformats-officedocument.presentationml.presentation":
		".pptx",
	"text/csv": ".csv",
	"text/plain": ".txt",
	"text/markdown": ".md",
};

/** Reverse map: extension to MIME */
const MIME_BY_EXT: Record<string, string> = {
	...Object.fromEntries(
		Object.entries(EXT_BY_MIME).map(([mime, ext]) => [ext, mime]),
	),
	".jpeg": "image/jpeg",
};

/** Audio file extensions */
const AUDIO_FILE_EXTENSIONS = new Set([
	".aac",
	".flac",
	".m4a",
	".mp3",
	".oga",
	".ogg",
	".opus",
	".wav",
]);

/** Voice-compatible audio extensions (Opus/Ogg) */
const VOICE_AUDIO_EXTENSIONS = new Set([".oga", ".ogg", ".opus"]);

/**
 * Normalize a MIME type from HTTP headers.
 */
function normalizeHeaderMime(mime?: string | null): string | undefined {
	if (!mime) return undefined;
	const cleaned = mime.split(";")[0]?.trim().toLowerCase();
	return cleaned || undefined;
}

/**
 * Detect MIME type from a buffer using magic bytes.
 */
async function sniffMime(
	buffer?: Buffer | Uint8Array,
): Promise<string | undefined> {
	if (!buffer) return undefined;
	try {
		const fileTypeFromBuffer = await getFileTypeFromBuffer();
		if (!fileTypeFromBuffer) return undefined;
		const type = await fileTypeFromBuffer(buffer);
		return type?.mime ?? undefined;
	} catch {
		return undefined;
	}
}

/**
 * Get the file extension from a path or URL.
 */
export function getFileExtension(filePath?: string | null): string | undefined {
	if (!filePath) return undefined;
	try {
		if (/^https?:\/\//i.test(filePath)) {
			const url = new URL(filePath);
			const ext = url.pathname.split(".").pop()?.toLowerCase();
			return ext ? `.${ext}` : undefined;
		}
	} catch {
		// fall back to plain path parsing
	}
	const parts = filePath.split(".");
	if (parts.length < 2) return undefined;
	return `.${parts.pop()?.toLowerCase()}`;
}

/**
 * Check if a MIME type is generic/container type.
 */
function isGenericMime(mime?: string): boolean {
	if (!mime) return true;
	const m = mime.toLowerCase();
	return m === "application/octet-stream" || m === "application/zip";
}

/**
 * Detect MIME type from buffer, headers, and/or file path.
 * Prioritizes sniffed types over extension-based detection.
 */
export async function detectMime(opts: {
	buffer?: Buffer | Uint8Array;
	headerMime?: string | null;
	filePath?: string;
}): Promise<string | undefined> {
	const ext = getFileExtension(opts.filePath);
	const extMime = ext ? MIME_BY_EXT[ext] : undefined;
	const headerMime = normalizeHeaderMime(opts.headerMime);
	const sniffed = await sniffMime(opts.buffer);

	// Prefer sniffed types, but don't let generic container types override
	// a more specific extension mapping (e.g. XLSX vs ZIP).
	if (sniffed && (!isGenericMime(sniffed) || !extMime)) {
		return sniffed;
	}
	if (extMime) return extMime;
	if (headerMime && !isGenericMime(headerMime)) return headerMime;
	if (sniffed) return sniffed;
	if (headerMime) return headerMime;

	return undefined;
}

/**
 * Get the file extension for a MIME type.
 */
export function extensionForMime(mime?: string | null): string | undefined {
	if (!mime) return undefined;
	return EXT_BY_MIME[mime.toLowerCase()];
}

/**
 * Check if a file appears to be an audio file by extension.
 */
export function isAudioFileName(fileName?: string | null): boolean {
	const ext = getFileExtension(fileName);
	return ext ? AUDIO_FILE_EXTENSIONS.has(ext) : false;
}

/**
 * Check if media is a GIF.
 */
export function isGifMedia(opts: {
	contentType?: string | null;
	fileName?: string | null;
}): boolean {
	if (opts.contentType?.toLowerCase() === "image/gif") return true;
	return getFileExtension(opts.fileName) === ".gif";
}

/**
 * Check if audio is voice-compatible (Opus/Ogg format).
 */
export function isVoiceCompatibleAudio(opts: {
	contentType?: string | null;
	fileName?: string | null;
}): boolean {
	const mime = opts.contentType?.toLowerCase();
	if (mime && (mime.includes("ogg") || mime.includes("opus"))) return true;
	const ext = getFileExtension(opts.fileName);
	return ext ? VOICE_AUDIO_EXTENSIONS.has(ext) : false;
}

/**
 * Get media kind from MIME type.
 */
export function mediaKindFromMime(mime?: string | null): MediaKind {
	if (!mime) return "unknown";
	const m = mime.toLowerCase();
	if (m.startsWith("image/")) return "image";
	if (m.startsWith("audio/")) return "audio";
	if (m.startsWith("video/")) return "video";
	if (
		m.startsWith("application/pdf") ||
		m.startsWith("application/msword") ||
		m.startsWith("application/vnd.ms-") ||
		m.startsWith("application/vnd.openxmlformats") ||
		m.startsWith("text/")
	) {
		return "document";
	}
	return "unknown";
}

/**
 * Get image MIME type from format name.
 */
export function imageMimeFromFormat(
	format?: string | null,
): string | undefined {
	if (!format) return undefined;
	switch (format.toLowerCase()) {
		case "jpg":
		case "jpeg":
			return "image/jpeg";
		case "heic":
			return "image/heic";
		case "heif":
			return "image/heif";
		case "png":
			return "image/png";
		case "webp":
			return "image/webp";
		case "gif":
			return "image/gif";
		default:
			return undefined;
	}
}

File: packages/rust/src/media/mime.rs
//! MIME type detection and media utilities for Eliza.
//!
//! Provides robust MIME type detection from file buffers, headers, and extensions.

use std::collections::HashMap;

/// Media kind categories
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum MediaKind {
    /// Image media (JPEG, PNG, GIF, etc.)
    Image,
    /// Audio media (MP3, OGG, WAV, etc.)
    Audio,
    /// Video media (MP4, MOV, etc.)
    Video,
    /// Document media (PDF, DOC, TXT, etc.)
    Document,
    /// Unknown or unrecognized media type
    Unknown,
}

impl std::fmt::Display for MediaKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            MediaKind::Image => write!(f, "image"),
            MediaKind::Audio => write!(f, "audio"),
            MediaKind::Video => write!(f, "video"),
            MediaKind::Document => write!(f, "document"),
            MediaKind::Unknown => write!(f, "unknown"),
        }
    }
}

lazy_static::lazy_static! {
    /// Map common MIME types to preferred file extensions
    static ref EXT_BY_MIME: HashMap<&'static str, &'static str> = {
        let mut m = HashMap::new();
        m.insert("image/heic", ".heic");
        m.insert("image/heif", ".heif");
        m.insert("image/jpeg", ".jpg");
        m.insert("image/png", ".png");
        m.insert("image/webp", ".webp");
        m.insert("image/gif", ".gif");
        m.insert("audio/ogg", ".ogg");
        m.insert("audio/mpeg", ".mp3");
        m.insert("audio/x-m4a", ".m4a");
        m.insert("audio/mp4", ".m4a");
        m.insert("video/mp4", ".mp4");
        m.insert("video/quicktime", ".mov");
        m.insert("application/pdf", ".pdf");
        m.insert("application/json", ".json");
        m.insert("application/zip", ".zip");
        m.insert("application/gzip", ".gz");
        m.insert("application/x-tar", ".tar");
        m.insert("application/x-7z-compressed", ".7z");
        m.insert("application/vnd.rar", ".rar");
        m.insert("application/msword", ".doc");
        m.insert("application/vnd.ms-excel", ".xls");
        m.insert("application/vnd.ms-powerpoint", ".ppt");
        m.insert("application/vnd.openxmlformats-officedocument.wordprocessingml.document", ".docx");
        m.insert("application/vnd.openxmlformats-officedocument.spreadsheetml.sheet", ".xlsx");
        m.insert("application/vnd.openxmlformats-officedocument.presentationml.presentation", ".pptx");
        m.insert("text/csv", ".csv");
        m.insert("text/plain", ".txt");
        m.insert("text/markdown", ".md");
        m
    };

    /// Reverse map: extension to MIME
    static ref MIME_BY_EXT: HashMap<&'static str, &'static str> = {
        let mut m: HashMap<&'static str, &'static str> = EXT_BY_MIME
            .iter()
            .map(|(k, v)| (*v, *k))
            .collect();
        m.insert(".jpeg", "image/jpeg");
        m
    };
}

/// Audio file extensions
const AUDIO_FILE_EXTENSIONS: &[&str] = &[
    ".aac", ".flac", ".m4a", ".mp3", ".oga", ".ogg", ".opus", ".wav",
];

/// Voice-compatible audio extensions (Opus/Ogg)
const VOICE_AUDIO_EXTENSIONS: &[&str] = &[".oga", ".ogg", ".opus"];

/// Normalize a MIME type from HTTP headers.
fn normalize_header_mime(mime: Option<&str>) -> Option<String> {
    mime.map(|m| m.split(';').next().unwrap_or("").trim().to_lowercase())
        .filter(|m| !m.is_empty())
}

/// Get the file extension from a path or URL.
pub fn get_file_extension(file_path: Option<&str>) -> Option<String> {
    let path = file_path?;

    // Try parsing as URL
    if path.to_lowercase().starts_with("http://") || path.to_lowercase().starts_with("https://") {
        if let Ok(url) = url::Url::parse(path) {
            let path_str = url.path();
            if let Some(ext) = path_str.rsplit('.').next() {
                if !ext.contains('/') {
                    return Some(format!(".{}", ext.to_lowercase()));
                }
            }
        }
    }

    // Plain path parsing
    let parts: Vec<&str> = path.split('.').collect();
    if parts.len() < 2 {
        return None;
    }
    Some(format!(".{}", parts.last()?.to_lowercase()))
}

/// Check if a MIME type is generic/container type.
fn is_generic_mime(mime: Option<&str>) -> bool {
    match mime {
        None => true,
        Some(m) => {
            let lower = m.to_lowercase();
            lower == "application/octet-stream" || lower == "application/zip"
        }
    }
}

/// Detect MIME type from buffer, headers, and/or file path.
/// Prioritizes sniffed types over extension-based detection.
pub fn detect_mime(
    buffer: Option<&[u8]>,
    header_mime: Option<&str>,
    file_path: Option<&str>,
) -> Option<String> {
    let ext = get_file_extension(file_path);
    let ext_mime = ext
        .as_ref()
        .and_then(|e| MIME_BY_EXT.get(e.as_str()).copied());
    let normalized_header = normalize_header_mime(header_mime);

    // Try to sniff MIME type from buffer using infer crate
    let sniffed = buffer.and_then(|b| infer::get(b).map(|kind| kind.mime_type().to_string()));

    // Prefer sniffed types, but don't let generic container types override
    // a more specific extension mapping (e.g., XLSX vs ZIP)
    if let Some(ref s) = sniffed {
        if !is_generic_mime(Some(s)) || ext_mime.is_none() {
            return Some(s.clone());
        }
    }
    if let Some(e) = ext_mime {
        return Some(e.to_string());
    }
    if let Some(ref h) = normalized_header {
        if !is_generic_mime(Some(h)) {
            return Some(h.clone());
        }
    }
    if sniffed.is_some() {
        return sniffed;
    }
    normalized_header
}

/// Get the file extension for a MIME type.
pub fn extension_for_mime(mime: Option<&str>) -> Option<&'static str> {
    mime.and_then(|m| EXT_BY_MIME.get(m.to_lowercase().as_str()).copied())
}

/// Check if a file appears to be an audio file by extension.
pub fn is_audio_filename(filename: Option<&str>) -> bool {
    get_file_extension(filename)
        .map(|ext| AUDIO_FILE_EXTENSIONS.contains(&ext.as_str()))
        .unwrap_or(false)
}

/// Check if media is a GIF.
pub fn is_gif_media(content_type: Option<&str>, filename: Option<&str>) -> bool {
    if let Some(ct) = content_type {
        if ct.to_lowercase() == "image/gif" {
            return true;
        }
    }
    get_file_extension(filename)
        .map(|ext| ext == ".gif")
        .unwrap_or(false)
}

/// Check if audio is voice-compatible (Opus/Ogg format).
pub fn is_voice_compatible_audio(content_type: Option<&str>, filename: Option<&str>) -> bool {
    if let Some(mime) = content_type {
        let lower = mime.to_lowercase();
        if lower.contains("ogg") || lower.contains("opus") {
            return true;
        }
    }
    get_file_extension(filename)
        .map(|ext| VOICE_AUDIO_EXTENSIONS.contains(&ext.as_str()))
        .unwrap_or(false)
}

/// Get media kind from MIME type.
pub fn media_kind_from_mime(mime: Option<&str>) -> MediaKind {
    match mime {
        None => MediaKind::Unknown,
        Some(m) => {
            let lower = m.to_lowercase();
            if lower.starts_with("image/") {
                MediaKind::Image
            } else if lower.starts_with("audio/") {
                MediaKind::Audio
            } else if lower.starts_with("video/") {
                MediaKind::Video
            } else if lower.starts_with("application/pdf")
                || lower.starts_with("application/msword")
                || lower.starts_with("application/vnd.ms-")
                || lower.starts_with("application/vnd.openxmlformats")
                || lower.starts_with("text/")
            {
                MediaKind::Document
            } else {
                MediaKind::Unknown
            }
        }
    }
}

/// Get image MIME type from format name.
pub fn image_mime_from_format(format_name: Option<&str>) -> Option<&'static str> {
    format_name.and_then(|fmt| match fmt.to_lowercase().as_str() {
        "jpg" | "jpeg" => Some("image/jpeg"),
        "heic" => Some("image/heic"),
        "heif" => Some("image/heif"),
        "png" => Some("image/png"),
        "webp" => Some("image/webp"),
        "gif" => Some("image/gif"),
        _ => None,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_get_file_extension() {
        assert_eq!(
            get_file_extension(Some("test.jpg")),
            Some(".jpg".to_string())
        );
        assert_eq!(
            get_file_extension(Some("path/to/file.PNG")),
            Some(".png".to_string())
        );
        assert_eq!(get_file_extension(Some("noext")), None);
        assert_eq!(get_file_extension(None), None);
    }

    #[test]
    fn test_media_kind_from_mime() {
        assert_eq!(media_kind_from_mime(Some("image/jpeg")), MediaKind::Image);
        assert_eq!(media_kind_from_mime(Some("audio/mp3")), MediaKind::Audio);
        assert_eq!(media_kind_from_mime(Some("video/mp4")), MediaKind::Video);
        assert_eq!(
            media_kind_from_mime(Some("application/pdf")),
            MediaKind::Document
        );
        assert_eq!(
            media_kind_from_mime(Some("text/plain")),
            MediaKind::Document
        );
        assert_eq!(
            media_kind_from_mime(Some("application/octet-stream")),
            MediaKind::Unknown
        );
        assert_eq!(media_kind_from_mime(None), MediaKind::Unknown);
    }

    #[test]
    fn test_is_audio_filename() {
        assert!(is_audio_filename(Some("song.mp3")));
        assert!(is_audio_filename(Some("voice.ogg")));
        assert!(!is_audio_filename(Some("image.jpg")));
        assert!(!is_audio_filename(None));
    }
}
